notebooks/adhoc/jovian_analysis/jovian_src/core.py

```diff
@@ -0,0 +1,1121 @@
+
+                        if progress_bar:
+                            progress_bar.close()
+
+                        if show_progress:
+                            success_count = len(dataset_paths) - error_count
+                            print(f"✅ Downloaded {success_count}/{len(dataset_paths)} files successfully")
+                            if error_count > 0:
+                                print(f"⚠️  {error_count} files had errors")
+
+            if not all_dfs:
+                if show_progress:
+                    print(f"❌ No transaction data found for {chain} on {date}")
+                return pl.DataFrame()
+
+            # Combine all dataframes
+            combined_df = pl.concat(all_dfs)
+
+            if show_progress:
+                print(f"✅ Successfully fetched {len(combined_df):,} transactions from GCS")
+                print(f"   Blocks: {combined_df['block_number'].n_unique():,}")
+                print(f"   Range: {combined_df['block_number'].min():,} - {combined_df['block_number'].max():,}")
+
+            return combined_df
+
+        except Exception as e:
+            if show_progress:
+                print(f"❌ Error fetching from GCS: {e}")
+                print(f"💡 Expected directory: {block_data_dir.absolute()}")
+            return pl.DataFrame()
+
+
+# =============================================================================
+# CALLDATA ANALYSIS
+# =============================================================================
+
+@dataclass
+class TransactionAnalysis:
+    """Result of analyzing a single transaction."""
+    transaction_hash: str
+    calldata_size: int
+    fastlz_size: int
+    size_estimate: float
+    calldata_footprint: float
+    compression_ratio: float
+
+
+@dataclass
+class BlockAnalysis:
+    """Result of analyzing a block."""
+    block_number: int
+    tx_count: int
+    total_calldata_footprint: float
+    total_size_estimate: float
+    calldata_utilization: float
+    exceeds_limit: bool
+    avg_footprint_per_tx: float
+    max_tx_footprint: float
+    transactions: List[TransactionAnalysis]
+    footprint_cost: int
+    block_gas_limit: int
+    block_gas_used: Optional[int] = None
+    utilization_vs_gas_used: Optional[float] = None
+    total_calldata_size: int = 0  # Sum of raw calldata sizes
+    total_fastlz_size: int = 0    # Sum of compressed sizes
+
+class CalldataAnalyzer:
+    """Core analyzer for Jovian calldata footprints."""
+
+    def __init__(self, jovian_config: JovianConfig, analysis_config: AnalysisConfig):
+        self.jovian_config = jovian_config
+        self.analysis_config = analysis_config
+        self.compressor = FastLZCompressor()
+
+    def calculate_size_estimate(self, calldata: bytes) -> float:
+        """Calculate Jovian size estimate for calldata."""
+        if not calldata:
+            return float(self.jovian_config.min_transaction_size)
+
+        fastlz_size = self.compressor.compress_len(calldata)
```

Comment:
We forgot to add a constant `68` overhead to account for transaction metadata like signatures.
```suggestion
        fastlz_size = self.compressor.compress_len(calldata) + 68
```
E.g. see the [Fjord specs](https://specs.optimism.io/protocol/fjord/predeploys.html) where fastLZ was introduced, where we also account for this in the L1 gas estimation functions by adding `68` to the fastLZ length of calldata.